docs/proposals/cascading-deletion.md

```diff
@@ -0,0 +1,157 @@
+
+## Expected race conditions
+* *Controller of the parent resource creates a child object after GC has observed the deletion of the parent object*: There are two possible cases and both are well handled in this design. 
+  * If parent’s orphaning policy is false, the parent object should have already been removed from the DAG, so when the child object is created, it’s added to the *Dirty Queue* and will be processed by the Garbage Processor.
+  * If parent’s orphaning policy is true, a tombstone should exist in the DAG, so the newly created child object will be added as child of the tombstone, and no further action will be taken.
+    * Note that this requires the tombstone to exist for a long enough period.
+
+* *GC deletes the child object before noticing the updates to the ParentReferences*: If this race turns out to be a problem in practice, the deletion request should be sent with ResouceVersion as a precondition. 
+
+
+# Design 2. Finalizer framework + GC
+
+There are two parts in this design. 2A is about the finalizer framework; 2B is about how to use the framework to enforce the orphaning policy.
+
+## 2A. Finalizer framework
+## API changes
+```
+type ObjectMeta struct {
+	…
+	DeletionInProgress bool
+}
+```
+**ObjectMeta.DeletionInProgress**: indicates if the object is in the progress of being deleted. It’s set by the API server. This field cannot be unset once set. Instead of adding this field, we can reuse ObjectMeta.DeletionTimestamp if it does not require the resource to be deleted after the time stamp is reached.
+
+```
+type ObjectMeta struct {
+	…
+	Finalizers map[string]string
+}
+```
+**ObjectMeta.Finalizers**: the map key is an ad hoc, explanatory name(e.g., "orphaning finalizer"), the value is an identifier for the responsible component that will remove the finalizer from the map. If the ObjectMeta.DeletionInProgress is true, entry in the map can only be removed.
+
+## New components
+* Finalizers:
+  * Like a controller, a finalizer is always running.
+  * Watches for update events that meet two conditions:
+    1. the updated objects has the identifier of the finalizer in ObjectMeta.Finalizers;
+    2. ObjectMeta.DeletionInProgress is updated from false to true.
+  * Applies the finalizing logic to the object in the update event.
+  * Removes itself from ObjectMeta.Finalizers after the finalizing logic is completed.
+  * Because it's possible for the finalizing logic to be applied multiple times, the finalizing logic has to be idempotent.
+
+## Changes to existing components
+* API server:
+  * Deletion handler:
+    * If the `ObjectMeta.Finalizers` of the object being deleted is non-empty, then updates the DeletionInProgress field to true.
+    * If the `ObjectMeta.Finalizers` is empty, then deletes the object.
+  * Update handler:
+    * If the update removes the last finalizer, and the DeletionInProgress is true, then deletes the object from the registry.
```

Comment:
If we reuse the DeletionTimestamp field, here API server also needs to check if DeletionGracePeriod is 0, if so, immediately deletes the object, otherwise just carries out the update.